File: daemons/macos/src/Handlers/PushHandler.swift
import Foundation

enum PushHandler {
    static func register(_ request: HTTPRequest) -> HTTPResponse {
        let body = (try? JSONSerialization.jsonObject(with: request.body)) as? [String: Any] ?? [:]
        let token = request.headers["x-push-device-token"] ?? body["token"] as? String
        if let deviceId = body["deviceId"] as? String, valid(deviceId, 128), let token,
            token.range(of: "^[A-Fa-f0-9]{32,512}$", options: String.CompareOptions.regularExpression) != nil,
            let environment = body["environment"] as? String, ["sandbox", "production"].contains(environment)
        {
            if PushDelivery.shared.available {
                PushDelivery.shared.register(deviceId: deviceId, token: token, environment: environment)
                return HTTPResponse.json(202, ["ok": true])
            }
            return HTTPResponse.json(503, ["error": "push_requires_provisioned_endpoint"])
        }
        return HTTPResponse.json(400, ["error": "invalid_push_device"])
    }

    private static func valid(_ value: String, _ maximum: Int) -> Bool {
        !value.isEmpty && value.count <= maximum
            && value.range(of: "^[A-Za-z0-9-]+$", options: String.CompareOptions.regularExpression) != nil
    }
}

File: daemons/linux/src/Handlers/PushHandler.js
import HTTPResponse from '../Networking/HTTPResponse.js'
import { pushDelivery } from '../Notifications/PushDelivery.js'

export function register(request) {
  const body = request.json()
  const token = request.headers['x-push-device-token'] || body.token
  if (typeof body.deviceId === 'string' && /^[A-Za-z0-9-]{1,128}$/u.test(body.deviceId) && typeof token === 'string' && /^[A-Fa-f0-9]{32,512}$/u.test(token) && ['sandbox', 'production'].includes(body.environment)) {
    const queued = pushDelivery.enqueue(`device:${body.deviceId}`, 'PUT', `/push-devices/${encodeURIComponent(body.deviceId)}`, { deviceId: body.deviceId, token, environment: body.environment, bundleId: 'soli.Cloude' })
    return HTTPResponse.json(queued ? 202 : 503, queued ? { ok: true } : { error: 'push_requires_provisioned_endpoint' })
  }
  return HTTPResponse.json(400, { error: 'invalid_push_device' })
}
